torch/_functorch/_aot_autograd/collect_metadata_analysis.py

```diff
@@ -57,7 +57,33 @@
 # In the future, you could imagine lifting this restriction, since these contiguous()
 # calls can have noticeable perf overhead depending on the model.
 def coerce_tangent(x):
-    return x.detach().contiguous() if isinstance(x, Tensor) else x
+    if not isinstance(x, Tensor):
+        return x
+    out = x.detach().contiguous()
+    # Note [Tangents must be contiguous, Part 2]
```

Comment:
cc @ezyang / @zou3519, I have CI mostly passing on my stack now so this PR is ready for another look. (I couldn't quite remember the final names we came up with from API bikeshedding Ed, but I tried to update them here).

My read of the situation of this PR is something like:

(1) these two new API's are not very ideal (two new subclass API's particular to tangents), but at the very least they are purely optional, you get a loud error in the rare situation that your subclass needed them but they weren't provided, and in the long-term state we can effectively forget about them

(2) the "right" solution would be to retrace the backward. There is still some design necessary, we should probably sit down with Horace and get in agreement on a long term design. It's probably not worth blocking internal models on this though, hence the PR.